### PR TITLE
Make iOS delegates opt-in

### LINKED ISSUE #39737: Provide way to exclude CoreML delegate from bazel build
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): master
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Currently when you build for iOS (specifically Swift in our case) you end up with the `coreml_delegate` target whether or not you consume it. As described in this comment, this may be something we want to support optionally excluding:

https://github.com/tensorflow/tensorflow/blob/d53c999feb2751b6a3ec9bace5be6f0b4620e4b4/tensorflow/lite/experimental/swift/BUILD.apple#L13-L24

**Will this change the current api? How?**

I think we could maintain the current behavior, and have a [`config_setting`](https://docs.bazel.build/versions/master/be/general.html#config_setting) that could be used with a `--define` such as `--define=EXCLUDE_COREML=1` that we could [`select`](https://docs.bazel.build/versions/master/be/functions.html#select) on to opt these out. This is done extensively in the [envoy build system](https://github.com/envoyproxy/envoy/blob/2b0633a09ce2ebf7ff4ba19ac470c013b1f2d35b/bazel/BUILD#L114) for similar reasons.

**Who will benefit with this feature?**

Anyone who doesn't need the CoreML specific logic and wants to potentially reduce their app size (in some cases this may have very little app size effect because of dead stripping)

**Any Other info.**

More discussion on https://github.com/tensorflow/tensorflow/pull/38956

### PR DESCRIPTION
**Pull request by keith**
Fixes https://github.com/tensorflow/tensorflow/issues/39737

### CODE REVIEW

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:32]: Would it be possible to do the selects here? I don't believe negating a select is possible, so if `use_coreml_delegate` is not true, then exclude `Sources/CoreMLDelegate.swift` as the "default" condition?

Same for Metal?

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:63]: Similar to comment for `TensorFlowLiteC_framework`, we will want Metal to follow same pattern as CoreML. I think @teijeong is planning to update Metal to match CoreML?

Probably out of scope for your current change, so seems fine to leave as you have it and we can add Metal to this list of deps when ready.

File: tensorflow/lite/experimental/ios/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/ios/BUILD.apple:20]: I think the plan is to move Metal delegate to its own framework, similar to CoreML delegate:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/ios/BUILD.apple#L67

In which case CoreML and Metal delegates will depend on `TensorFlowLiteC_framework`. So the Metal delegate selects in this target will eventually be moved.

Since this is probably out of scope for your change, could probably leave it as you have it and we can migrate metal delegate to match CoreML delegate in follow-up change.

File: tensorflow/lite/tools/delegates/BUILD
Comment (tristane0) [inline tensorflow/lite/tools/delegates/BUILD:24]: From my understanding, CoreML and Metal are two separate delegates that are independent of each other. I think once Metal delegate build config is updated to match current CoreML delegate config that it should be a bit more clear.

**Review by ghost — COMMENTED** [review]
This is great! Thank you @keith for creating the PR!

Left a few comments regarding current state for CoreML vs Metal delegates, which is a bit confusing at the moment since we are still in the process of updating Metal delegate build config to match CoreML. 

Also, do you mind adding some example invocations that would demonstrate how to include 1 or both of the delegates?

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (keith) [reply tensorflow/lite/experimental/swift/BUILD.apple:32]: I tried something like that (if I understand what you mean) but the value of `exclude` has to be a list of strings, it can't be a select, which was a TIL

File: tensorflow/lite/tools/delegates/BUILD
Comment (keith) [reply tensorflow/lite/tools/delegates/BUILD:24]: Hrm, the reason I combined these was because I hit some link errors when only using the CoreML one without the Metal one, but maybe there are some missing dependencies elsewhere that need to be fixed instead:

```
Undefined symbols for architecture x86_64:
  "tflite::delegates::GraphPartitionHelper::Partition(std::__1::set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >*)", referenced from:
      tflite::(anonymous namespace)::DelegatePrepare(TfLiteContext*, TfLiteDelegate*) in libcoreml_delegate.a(coreml_delegate_5cb2e341c6e1d2730c9dcde9cd76c559.o)
  "tflite::delegates::GraphPartitionHelper::GetFirstNLargestPartitions(int, int) const", referenced from:
      tflite::(anonymous namespace)::DelegatePrepare(TfLiteContext*, TfLiteDelegate*) in libcoreml_delegate.a(coreml_delegate_5cb2e341c6e1d2730c9dcde9cd76c559.o)
  "vtable for tflite::delegates::GraphPartitionHelper", referenced from:
      tflite::delegates::GraphPartitionHelper::GraphPartitionHelper(TfLiteContext*, std::__1::function<bool (TfLiteContext*, TfLiteNode*, TfLiteRegistration*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)>) in libcoreml_delegate.a(coreml_delegate_5cb2e341c6e1d2730c9dcde9cd76c559.o)
      tflite::delegates::GraphPartitionHelper::~GraphPartitionHelper() in libcoreml_delegate.a(coreml_delegate_5cb2e341c6e1d2730c9dcde9cd76c559.o)
```

**Comment by keith** [comment]
Added some invocation examples as comments, lmk if you had somewhere else in mind!

**Comment by keith** [comment]
I'll actually have to add another example after https://github.com/tensorflow/tensorflow/pull/38956 merges

File: tensorflow/lite/tools/delegates/BUILD
Comment (tristane0) [reply tensorflow/lite/tools/delegates/BUILD:24]: This may need to be updated as part of refactor for separating Metal delegate from the C library. 
@teijeong to confirm

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [reply tensorflow/lite/experimental/swift/BUILD.apple:32]: I see, was thinking something like:
```
    srcs = glob(
        [
            "Sources/*.swift",
        ],
        exclude = select({
          "//tensorflow/lite/tools/delegates:use_coreml_delegate": [],
          "//conditions:default": [
              "Sources/CoreMLDelegate.swift",
          ],
        }) + select({
          "//tensorflow/lite/tools/delegates:use_metal_delegate": [],
          "//conditions:default": [
              "Sources/MetalDelegate.swift",
          ],
        }),
    )
```

OR

```
    srcs = glob(
        [
            "Sources/*.swift",
        ],
        exclude = [] + select({
          "//tensorflow/lite/tools/delegates:use_coreml_delegate": [],
          "//conditions:default": [
              "Sources/CoreMLDelegate.swift",
          ],
        }) + select({
          "//tensorflow/lite/tools/delegates:use_metal_delegate": [],
          "//conditions:default": [
              "Sources/MetalDelegate.swift",
          ],
        }),
    )
```

But if that's less readable and still doesn't work, then I think what you have is reasonable. Was just looking for a way to avoid having to update files / selects in multiple places when adding new delegates in future.

Comment (keith) [reply tensorflow/lite/experimental/swift/BUILD.apple:32]: Yea we can't put the `select` in the `exclude` so this is what we're stuck with 😢

**Comment by tristane0** [comment]
Thanks again for your help! Changes look good to me.

Will let @teijeong take a final pass, when he has a chance.

File: tensorflow/lite/experimental/delegates/coreml/BUILD
Comment (keith) [inline tensorflow/lite/experimental/delegates/coreml/BUILD:59]: This was the missing dependency that made it seem that the CoreML delegate depended on the Metal delegate

File: tensorflow/lite/experimental/ios/BUILD.apple
Comment (keith) [inline tensorflow/lite/experimental/ios/BUILD.apple:20]: This change now just removes the metal header, since that should only come now if you use the metal framework

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (keith) [reply tensorflow/lite/experimental/swift/BUILD.apple:63]: The metal cocoapods changes were merged yesterday, I've rebased to work with the new pattern

**Review by ghost — DISMISSED** [review]
Thanks @keith! Updates look good to me.

@yyoon / @teijeong can either of you merge this PR once #38956 has been merged? Thank you!

File: tensorflow/lite/experimental/delegates/coreml/BUILD
Comment (teijeong) [reply tensorflow/lite/experimental/delegates/coreml/BUILD:59]: Oops I missed this while exporting to github. Thanks for catching!

**Review by yyoon — COMMENTED** [review]
There's one issue that concerns me, which I believe requires some more discussion.

Let's say I'm developing an iOS app with bazel, and always want to include Metal delegate. To make the final `ios_application` be built correctly with the metal delegate dependency, I would have two choices.

1) Require everyone to build the app with `--define use_metal_delegate=1` option from the command line.
2) Create my own `swift_library` target with the metal delegate sources included and use it instead.

I think neither is really acceptable here. Ideally, anyone should be able to build my `ios_application` target without necessarily adding the `--define` flag. Also, having to create my own `swift_library` target feels like an unnecessary burden, and may not be even possible if the visibility of the source files are tightly managed.

Have you thought about this issue? Do you have better suggestions for the app developers wanting to use the `TensorFlowLite` swift library? How is your team going to handle this situation?

File: tensorflow/lite/tools/delegates/BUILD
Comment (keith) [reply tensorflow/lite/tools/delegates/BUILD:12]: I wondered, thanks, done

File: tensorflow/lite/delegates/BUILD
Comment (tristane0) [inline tensorflow/lite/delegates/BUILD:23]: nit: Do we want to add a comment for each of these `config_setting`s to indicate what they enable and when they should be used, i.e. for building `tensorflow_lite_c` (cc_library) and `TensorFlowLite` (swift_library)?

**Comment by keith** [comment]
> Have you thought about this issue?

This is definitely a good callout. 

> How is your team going to handle this situation?

After a few years of using bazel we have found there is absolutely no way to be successful without a significant amount of configuration in a root `.bazelrc` file, so I didn't even consider this as an issue since I assume any significantly complex iOS application will have to have the same. If that's true the burden of folks including that for their entire team would be very low (this is what we would do if we were planning on using the delegates, we actually want the codepath with none of them at least for now)

> Do you have better suggestions for the app developers wanting to use the TensorFlowLite swift library? 

Off the top of my head I think we have 2 options:

1. Switch to the blacklist approach. In this case you'd have the same issue, but with reverse defines, but maybe this would be good enough
2. Create a new `swift_library` target that is virtually a duplicate of the current one except with none of the `select()` statements, so folks could depend on that directly for "all" the delegates. This would also only partially solve the issue if you wanted only 1 of them.

In general I think this just brings up a place where bazel lacks some configurability, and flags are our only option 😢

File: tensorflow/lite/delegates/BUILD
Comment (keith) [reply tensorflow/lite/delegates/BUILD]: So now that these actually aren't used in the iOS BUILD file, I've moved them directly into the Swift BUILD file, lmk what you think

Comment (tristane0) [reply tensorflow/lite/delegates/BUILD]: That's a good point! I think the plan may be to add delegation support to `//tensorflow/lite/experimental/objc:TensorFlowLite` in the future, in which case it would probably be better to move the configs to `//tensorflow/lite/experimental/ios/BUILD` so it can be shared between Swift and ObjC. 

@teijeong / @yyoon to confirm.

Comment (keith) [reply tensorflow/lite/delegates/BUILD]: Happy to move it now, or we can as needed

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:23]: I think we can remove this TODO to avoid confusion as your PR is addressing the issue.

Comment (keith) [reply tensorflow/lite/experimental/swift/BUILD.apple:23]: Done!

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:36]: We may want to add a comment stating that by default this target builds the core TensorFlow Lite Swift library and that developers will need to specify the appropriate defines to enable delegation support.

Comment (keith) [reply tensorflow/lite/experimental/swift/BUILD.apple:23]: Moved the other comment up here since it makes more sense to be with this I think

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:89]: remove?

Comment (keith) [reply tensorflow/lite/experimental/swift/BUILD.apple]: ah sorry, it finally snuck in, fixed

**Comment by yyoon** [comment]
> After a few years of using bazel we have found there is absolutely no way to be successful without a significant amount of configuration in a root `.bazelrc` file

This is a fair point, and I agree that lots of projects require additional parameters anyways.

My concern was also related to how our internal CI system works. There are existing apps already depending on the `TensorFlowLite` Swift library, and because some of them are using Metal delegate, those builds would probably break if we submitted this PR as is. There are ways to provide custom build flags for each project, but it's a bit more involved than it sounds.

> 1. Switch to the blacklist approach. In this case you'd have the same issue, but with reverse defines, but maybe this would be good enough

Yes, I would actually much prefer an opt-out model, because it could still give you options to easily exclude unnecessary delegates, and when we don't really care about the project size, the default configuration would suffice. I believe this would be also good enough to meet your needs.

More specifically, we could rename the define variables to be `use_foo_delegate` -> `exclude_foo_delegate`, and reverse the select conditions.

**Comment by tristane0** [comment]
> Yes, I would actually much prefer an opt-out model, because it could still give you options to easily exclude unnecessary delegates, and when we don't really care about the project size, the default configuration would suffice. I believe this would be also good enough to meet your needs.

The current CocoaPods approach is to opt-in, where developers explicitly list in their Podfiles which delegates they want to include. Do we want Bazel to match that workflow?

> There are existing apps already depending on the TensorFlow Lite Swift library, and because some of them are using Metal delegate, those builds would probably break if we submitted this PR as is. 

Agreed that this would need to be communicated to existing teams, but you can think about the reverse where that each time we add a new delegate to `TensorFlowLite` Swift library, teams will need to decide if they want to increase their binary size for one or more delegates that they don't plan to use. Opt-in approach breaks existing teams currently using Metal delegation, but hopefully long term will provide them with the flexibility to add only the delegates they plan to use and avoid unnecessary binary increases. 

If the plan is to add more delegates in the future, I think opt-in for both CocoaPods and Bazel provides the most flexibility for consumers of the TensorFlow Lite Swift library.

**Comment by yyoon** [comment]
@tristane0 Okay, that makes sense to me. I'm convinced enough to proceed with opt-in approach. 😀

**Review by yyoon — DISMISSED** [review]

**Comment by tristane0** [comment]
@keith looks like this change will break the existing unit tests for [MetalDelegate](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/swift/Tests/MetalDelegateTests.swift)

We may need to exclude `MetalDelegateTests.swift` and do the selects here as well:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/swift/BUILD.apple#L70

and add the `use_metal_delegate=1` define to where we invoke `bazel test //tensorflow/lite/experimental/swift/...`

Any other options?

**Comment by keith** [comment]
I pushed what I think is the other option, which might be a bit nicer overall, lmk what you think or I can revert it and go down the other path. I think it kinda depends on how much of a pain it will be to keep the second one in sync, vs how much of a pain it'll be to run CI multiple times with different defines. I wish there was a way to force the defines on from the `swift_library` level or `select` on `features` but I can't think of another way to configure it.

**Comment by tristane0** [comment]
> I pushed what I think is the other option, which might be a bit nicer overall, lmk what you think or I can revert it and go down the other path. I think it kinda depends on how much of a pain it will be to keep the second one in sync, vs how much of a pain it'll be to run CI multiple times with different defines. I wish there was a way to force the defines on from the `swift_library` level or `select` on `features` but I can't think of another way to configure it.

Thank you @keith! This was one of the options we were discussing yesterday and we agreed that it was a simpler approach than dealing with the selects for the `Tests` library. I am fine with moving forward with this approach. 

Would it be possible to mark the "all delegates" `swift_library` as `testonly = 1`? I think this would also require updating the test app targets to include that as well? I would also be ok with having test app just depend on `:TensorFlowLite` as it doesn't reference any of the delegates and not sure if there are any plans to do so since TFLite has existing [example apps](https://github.com/tensorflow/examples/tree/master/lite/examples) that already demonstrate Metal delegate usage.

@yyoon to confirm and approve.

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:79]: can we mark this target as `test_only`?

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:76]: an->and?

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:156]: Would be fine with reverting this change as the test app doesn't currently reference any delegates and I don't think there are plans to due to.

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:76]: If we remove from the test app below, then I think we can just say that this target is for tests only, or allow the `test_only=1` config (if set) to provide implicit documentation for the target.

**Comment by keith** [comment]
Updated!

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (tristane0) [inline tensorflow/lite/experimental/swift/BUILD.apple:155]: `:TensorFlowLite`?

**Comment by tristane0** [comment]
> Updated!

Looks like we need to update the `:TestsLibrary` to depend on the "all delegates" target; otherwise, LGTM. Thank you!

**Comment by keith** [comment]
Sorry I swapped the wrong one, updated

**Review by ghost — DISMISSED** [review]

**Comment by rthadur** [comment]
@keith can you please check sanity errors [here](https://source.cloud.google.com/results/invocations/cac62d44-f0db-4637-8652-c431f800792b/log)

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (yyoon) [inline tensorflow/lite/experimental/swift/BUILD.apple:77]:
```suggestion
    name = "TensorFlowLiteAllDelegates",
    testonly = 1,
```

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (yyoon) [inline tensorflow/lite/experimental/swift/BUILD.apple:84]:
```suggestion
```

File: tensorflow/lite/experimental/swift/BUILD.apple
Comment (yyoon) [inline tensorflow/lite/experimental/swift/BUILD.apple:89]:
```suggestion
        "//tensorflow/lite/delegates/gpu:metal_delegate",
        "//tensorflow/lite/experimental/delegates/coreml:coreml_delegate",
```

**Review by yyoon — COMMENTED** [review]
I believe these suggested changes should fix the sanity check.

**Comment by keith** [comment]
thanks, ran buildifier on it locally

**Review by yyoon — APPROVED** [review]
Sounds good. Thanks!

**Pull request merged**
